File: monsoon3.0/src/main/java/wtf/monsoon/impl/module/player/InventoryManager.java
package wtf.monsoon.impl.module.player;

import io.github.nevalackin.homoBus.Listener;
import io.github.nevalackin.homoBus.annotations.EventLink;
import net.minecraft.block.Block;
import net.minecraft.client.gui.inventory.GuiContainer;
import net.minecraft.client.gui.inventory.GuiInventory;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.init.Blocks;
import net.minecraft.inventory.Slot;
import net.minecraft.item.*;
import net.minecraft.network.play.client.C0DPacketCloseWindow;
import net.minecraft.network.play.client.C0EPacketClickWindow;
import net.minecraft.network.play.client.C16PacketClientStatus;
import net.minecraft.potion.PotionEffect;
import wtf.monsoon.Wrapper;
import wtf.monsoon.api.module.Category;
import wtf.monsoon.api.module.Module;
import wtf.monsoon.api.setting.Setting;
import wtf.monsoon.api.util.entity.MovementUtil;
import wtf.monsoon.api.util.entity.PlayerUtil;
import wtf.monsoon.api.util.misc.PacketUtil;
import wtf.monsoon.api.util.misc.Timer;
import wtf.monsoon.impl.event.EventPacket;
import wtf.monsoon.impl.event.EventUpdate;
import wtf.monsoon.impl.module.combat.AutoPot;

import java.util.*;

public class InventoryManager extends Module {

    private final Timer timer = new Timer();

    private int lastSlot;

    public Setting<Double> delay = new Setting<>("Delay", 50.0)
            .minimum(0D)
            .maximum(500D)
            .incrementation(5D)
            .describedBy("The delay between each inventory update.");

    public Setting<Boolean> open = new Setting<>("Open Inventory", false)
            .describedBy("Open the inventory.");

    public Setting<Boolean> spoof = new Setting<>("Spoof Inventory", false)
            .describedBy("Spoof having your inventory open server-side. Can cause desyncing.");

    public Setting<Boolean> whileNotMoving = new Setting<>("While not Moving", false)
            .describedBy("Only manage your inventory if you aren't moving.");

    public Setting<Boolean> moveBlocksToHotbar = new Setting<>("Move blocks to hotbar", true)
            .describedBy("Whether to move blocks to your hotbar.");

    public Setting<Boolean> movePotionsToHotbar = new Setting<>("Move potions to hotbar", true)
            .describedBy("Whether to move potions to your hotbar.");

    public List<String> junk = Arrays.asList("stick", "egg", "string", "cake", "mushroom", "flint", "dyePowder", "feather", "chest", "snowball", "fish", "enchant", "exp", "shears", "anvil", "torch", "seeds", "leather", "reeds", "skull", "record", "piston", "snow", "bottle", "poison", "shield", "web", "chest", "bucket");

    private boolean inventoryOpen = false;

    public InventoryManager() {
        super("Manager", "Manages your Inventory", Category.PLAYER);
    }

    @Override
    public void onEnable() {
        super.onEnable();
        timer.reset();
        lastSlot = 0;
    }

    @Override
    public void onDisable() {
        super.onDisable();
        lastSlot = 0;
    }

    @EventLink
    public final Listener<EventUpdate> eventUpdateListener = e -> {

        if (timer.hasTimeElapsed((long) delay.getValue().doubleValue(), true) /*&& !Monsoon.INSTANCE.manager.getModuleByClass(AutoArmor.class).isRunning*/) {
            if (isInventoryOpen() && (!player.isMoving() || !whileNotMoving.getValue())) {
                if ((mc.currentScreen == null || (mc.currentScreen instanceof GuiContainer && ((GuiContainer) mc.currentScreen).inventorySlots == mc.thePlayer.inventoryContainer))) {
                    for (int i = 9; i < 45; i++) {
                        if (lastSlot >= 45) lastSlot = 0;
                        while (i <= lastSlot) i++;
                        lastSlot = i;
                        Slot slot;
                        try {
                            slot = mc.thePlayer.inventoryContainer.getSlot(i);
                        } catch (Exception ex) {
                            continue;
                        }

                        if (!slot.getHasStack()) continue;

                        ItemStack stack = slot.getStack();
                        Item item = stack.getItem();

                        if (item instanceof ItemSword && isBestSword(i)) {
                            openInv();
                            mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, i, 0, 2, mc.thePlayer);
                            closeInv();
                            return;
                        } else if (item instanceof ItemSword && !isBestSword(i)) {
                            dropItem(i);
                        }

                        if (item instanceof ItemTool && !isBestTool(i)) {
                            //Monsoon.sendMessage("not best tool " + stack.getDisplayName());
                            dropItem(i);
                            return;
                        }

                        if (isJunk(i) && !(item instanceof ItemArmor)) {
                            // Monsoon.sendMessage("junk " + stack.getDisplayName());
                            dropItem(i);
                            return;
                        }

                        if(item instanceof ItemPotion && movePotionsToHotbar.getValue()) {

                            ItemPotion potion = (ItemPotion) item;

                            if(!ItemPotion.isSplash(stack.getMetadata())) continue;

                            boolean shouldMove = false;

                            for(PotionEffect potionEffect : potion.getEffects(stack.getMetadata())) {
                                if(Wrapper.getModule(AutoPot.class).isValidEffect(potionEffect)) shouldMove = true;
                            }

                            if(shouldMove) {
                                openInv();
                                mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, i, 0, 1, mc.thePlayer);
                                closeInv();
                            }
                            return;
                        }

                        int blocksInHotbar = 0;

                        for (int j = 0; j < 9; j++) {
                            if(blocksInHotbar > 128) continue;
                            try {
                                if (!mc.thePlayer.inventoryContainer.getSlot(j + 36).getHasStack()) continue;
                                if(mc.thePlayer.inventoryContainer.getSlot(j + 36).getStack().getItem() instanceof ItemBlock) {
                                    blocksInHotbar += mc.thePlayer.inventoryContainer.getSlot(j + 36).getStack().stackSize;
                                }
                            } catch (Exception ignored) {
                            }
                        }

                        if(item instanceof ItemBlock && stack.stackSize > 8 && i < 36 && moveBlocksToHotbar.getValue() && blocksInHotbar < 96)  {
                            openInv();
                            mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, i, 0, 1, mc.thePlayer);
                            closeInv();
                            return;
                        }

                        if(item.getItemStackLimit() > 1 && i < 36) {
                            for (int j = 0; j < 9; j++) {
                                try {
                                    if (!mc.thePlayer.inventoryContainer.getSlot(j + 36).getHasStack()) continue;
                                    ItemStack jStack = mc.thePlayer.inventoryContainer.getSlot(j + 36).getStack();
                                    Item jItem = jStack.getItem();
                                    if(jItem.getClass().equals(item.getClass()) && jStack.stackSize < 64) {
                                        openInv();
                                        mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, i, 0, 1, mc.thePlayer);
                                        closeInv();
                                        return;
                                    }
                                } catch (Exception ignored) {
                                }
                            }
                        }
                    }
                }
            } else {
                lastSlot = 0;
            }
        }
    };

    private boolean isJunk(int slotid) {
        Slot slot = mc.thePlayer.inventoryContainer.getSlot(slotid);
        ItemStack stack = slot.getStack();

        if (stack != null) {
            Item item = stack.getItem();
            if (item instanceof ItemBanner) return true;
            if (stack.getDisplayName().toLowerCase().contains("tnt")) return true;
            if (stack.getDisplayName().toLowerCase().contains("bucket")) return true;
            if (stack.getDisplayName().toLowerCase().contains("bottle")) return true;
            for (String shortName : junk) {
                if (stack.getDisplayName().toLowerCase().contains(shortName)) return true;
                if (item.getUnlocalizedName().toLowerCase().contains(shortName)) return true;
            }
        }

        return false;
    }

    private boolean isBestTool(int slotid) {
        Slot slot = mc.thePlayer.inventoryContainer.getSlot(slotid);
        ItemStack compareStack = slot.getStack();

        for (int i = 9; i < 45; i++) {
            Slot slot2 = mc.thePlayer.inventoryContainer.getSlot(i);
            ItemStack stack2 = slot2.getStack();

            if (stack2 != null && compareStack != stack2 && stack2.getItem() instanceof ItemTool) {
                ItemTool item = (ItemTool) stack2.getItem();
                ItemTool compare = (ItemTool) compareStack.getItem();
                if (item.getClass() == compare.getClass()) {
                    if (compare.getStrVsBlock(stack2, preferredBlock(item.getClass())) <= item.getStrVsBlock(compareStack, preferredBlock(compare.getClass())))
                        return false;
                }
            }
        }

        return true;
    }

    private boolean isBestSword(int slotid) {
        Slot slot = mc.thePlayer.inventoryContainer.getSlot(slotid);
        ItemStack compareStack = slot.getStack();
        for (int i = 9; i < 45; i++) {
            Slot slot2 = mc.thePlayer.inventoryContainer.getSlot(i);
            ItemStack stack = slot2.getStack();

            if (stack != null && compareStack != stack && stack.getItem() instanceof ItemSword) {
                if (compareStack.getItem() instanceof ItemSword) {
                    ItemSword item = (ItemSword) stack.getItem();
                    ItemSword compare = (ItemSword) compareStack.getItem();
                    if (item.getClass() == compare.getClass()) {
                        if (compare.attackDamage + getSwordStrength(compareStack) <= item.attackDamage + getSwordStrength(stack))
                            return false;
                    }
                }
            }
        }

        return true;
    }

    private boolean dropItem(int slot) {
        openInv();
        Slot slo2t = mc.thePlayer.inventoryContainer.getSlot(slot);
        ItemStack stack = slo2t.getStack();
        // Monsoon.sendMessage("dropped " + stack.getDisplayName());
        mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, slot, 1, 4, mc.thePlayer);
        closeInv();
        return true;
    }

    private Block preferredBlock(Class clazz) {
        return clazz == ItemPickaxe.class ? Blocks.cobblestone : clazz == ItemAxe.class ? Blocks.log : Blocks.dirt;
    }

    private float getSwordStrength(ItemStack stack) {
        return (!(stack.getItem() instanceof ItemSword) ? 0.0F : (float) EnchantmentHelper.getEnchantmentLevel(Enchantment.sharpness.effectId, stack) * 1.25F) + (!(stack.getItem() instanceof ItemSword) ? 0.0F : (float) EnchantmentHelper.getEnchantmentLevel(Enchantment.fireAspect.effectId, stack));
    }

    private int checkProtection(ItemStack item) {
        return EnchantmentHelper.getEnchantmentLevel(0, item);
    }

    @EventLink
    public final Listener<EventPacket> eventPacketListener = e -> {
        if(e.getPacket() instanceof C16PacketClientStatus) {
            C16PacketClientStatus packet = (C16PacketClientStatus) e.getPacket();
            if(packet.getStatus().equals(C16PacketClientStatus.EnumState.OPEN_INVENTORY_ACHIEVEMENT)) {
                inventoryOpen = true;
            }
        }

        if(e.getPacket() instanceof C0DPacketCloseWindow) {
            C0DPacketCloseWindow packet = (C0DPacketCloseWindow) e.getPacket();
            if(packet.getWindowId() == mc.thePlayer.inventoryContainer.windowId) {
                inventoryOpen = false;
            }
        }
    };

    private void openInv() {
        if (!inventoryOpen && spoof.getValue()) PacketUtil.sendPacketNoEvent(new C16PacketClientStatus(C16PacketClientStatus.EnumState.OPEN_INVENTORY_ACHIEVEMENT));
        inventoryOpen = true;
    }

    private void closeInv() {
        if (inventoryOpen && spoof.getValue()) PacketUtil.sendPacketNoEvent(new C0DPacketCloseWindow(mc.thePlayer.inventoryContainer.windowId));
        inventoryOpen = false;
    }

    private boolean isInventoryOpen() {
        if(open.getValue()) return mc.currentScreen instanceof GuiInventory;
        //else if(spoof.getValue()) return this.inventoryOpen;
        else return true;
    }

}

File: monsoon-3.x-v2/src/main/java/wtf/monsoon/client/modules/player/InventoryManager.kt
package wtf.monsoon.client.modules.player

import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.client.gui.inventory.GuiContainerCreative
import net.minecraft.client.gui.inventory.GuiInventory
import net.minecraft.inventory.Slot
import net.minecraft.item.*
import wtf.monsoon.backend.Category
import wtf.monsoon.backend.module.Module
import wtf.monsoon.backend.setting.Setting
import wtf.monsoon.client.event.EventPacket
import wtf.monsoon.client.event.EventPreMotion
import wtf.monsoon.client.event.EventRender2D
import wtf.monsoon.client.util.math.randomNumber
import wtf.monsoon.client.util.misc.Stopwatch
import wtf.monsoon.client.util.player.ItemUtil
import java.awt.Color

class InventoryManager : Module("Manager", "Manages your inventory.", Category.PLAYER) {

    private val delay = Setting("Delay", "The delay to use", 50L, 0L, 1000L, 10L)

    private val enableRandomization = Setting("Randomization", true)
    private val randomDelay = Setting("Range", "Range of randomization", 10L, 1L, 50L, 1L)
        .visibleWhen { enableRandomization.getValue() }

    private val openInv = Setting("While Inventory Open", false)
    private val whileNotMoving = Setting("While not moving", false)

    private val maxBlocks = Setting("Maximum Blocks", "The maximum amount of blocks to allow in your inventory.", 128, 32, 512, 8)

    private val slotsGroup = Setting<String>("Slots", "container")
    private val swordSlot = Setting("Sword Slot", "Slot to place your sword in", 1, 0, 9, 1).childOf(slotsGroup)
    private val pickaxeSlot = Setting("Pickaxe Slot", "Slot to place your pickaxe in", 2, 0, 9, 1).childOf(slotsGroup)
    private val axeSlot = Setting("Axe Slot", "Slot to place your axe in", 3, 0, 9, 1).childOf(slotsGroup)
    private val shovelSlot = Setting("Shovel Slot", "Slot to place your shovel in", 4, 0, 9, 1).childOf(slotsGroup)
    private val blockSlot = Setting("Blocks Slot", "Slot to place your blocks in", 5, 0, 9, 1).childOf(slotsGroup)
    private val gappleSlot = Setting("Golden Apple Slot", "Slot to place your gaps in", 6, 0, 9, 1).childOf(slotsGroup)

    private val slots = arrayOf(swordSlot, pickaxeSlot, axeSlot, shovelSlot)

    var timer = Stopwatch()
    var time = this.delay.getValue()
    var index = 0
    var lastIndex = 0

    override fun enable() {
        super.enable()
        this.timer.reset()
        this.time = this.delay.getValue()
        this.index = 9
        this.lastIndex = 9
    }

    @EventListener
    val preMotion = fun(_: EventPreMotion) {
        this.manage()
    }

    private fun manage() {
        if(this.timer.hasTimeElapsed(this.time)) {
            if(this.canRun() && this.slotsAreValid()) {
                val inventory = mc.thePlayer.inventoryContainer

                if(this.index == this.lastIndex)
                    this.index++

                if(this.index >= inventory.inventorySlots.size)
                    this.index = 9

                this.lastIndex = this.index

                if(!inventory.getSlot(index).hasStack)
                    return

                val stack: ItemStack = inventory.getSlot(index).stack!!
                val item: Item = stack.item

                if (item is ItemSword || item is ItemArmor || item is ItemTool) {
                    if(this.isDuplicate(this.index, stack, item))
                        return this.dropItem(this.index, inventory.windowId)

                    if (item is ItemSword)
                        if (ItemUtil.isBestSword(stack, item))
                            if(this.moveToSlot(this.index, this.swordSlot.getValue() - 1, inventory.windowId)) return
                        else
                            return this.dropItem(this.index, inventory.windowId)

                    if (item is ItemArmor)
                        if (!ItemUtil.isBestArmor(stack, item))
                            return this.dropItem(this.index, inventory.windowId)
                        else
                            if(this.moveToSlot(this.index, this.getArmorSlot(item), inventory.windowId)) return

                    if (item is ItemTool)
                        if (!ItemUtil.isBestTool(stack, item))
                            return this.dropItem(this.index, inventory.windowId)
                        else
                            if(this.moveToSlot(this.index, this.getToolSlot(item.javaClass) - 1, inventory.windowId)) return
                } else {
                    if (item is ItemBlock) {
                        if(this.getBlocksInInventory() > this.maxBlocks.getValue()) {
                            return this.dropItem(this.index, inventory.windowId)
                        }

                        if(!inventory.getSlot(this.blockSlot.getValue()).hasStack) {
                            if(this.moveToSlot(this.index, this.blockSlot.getValue() - 1, inventory.windowId)) return
                        } else if(inventory.getSlot(this.blockSlot.getValue()).stack.item !is ItemBlock) {
                            if(this.moveToSlot(this.index, this.blockSlot.getValue() - 1, inventory.windowId)) return
                        }
                    }

                    if (item is ItemAppleGold)
                        if(this.moveToSlot(this.index, this.gappleSlot.getValue() - 1, inventory.windowId)) return


                    if (ItemUtil.isJunk(stack, item))
                        return this.dropItem(this.index, inventory.windowId)
                }
            } else
                this.index = 9
        }
    }

    @EventListener
    val render2D = fun(e: EventRender2D) {
        if(!this.slotsAreValid()) {
            val fr = mc.fontRendererObj
            val message = "Two or more slots are the same in Manager settings, the module will not function until this is fixed."
            val l1: Int = e.scaledResolution.scaledWidth / 2 - fr.getStringWidth(message) / 2
            val i1: Int = e.scaledResolution.scaledHeight / 2 + 8
            fr.drawString(message, l1 + 1, i1, 0)
            fr.drawString(message, l1 - 1, i1, 0)
            fr.drawString(message, l1, i1 + 1, 0)
            fr.drawString(message, l1, i1 - 1, 0)
            fr.drawString(message, l1, i1, Color.RED.rgb)
        }
    }

    private fun isDuplicate(slot: Int, stack: ItemStack, item: Item): Boolean {
        if(item is ItemTool) {
            for (i in 0..44) {
                if(i == slot) continue
                val slot2: Slot = ItemUtil.mc.thePlayer.inventoryContainer.getSlot(i)
                if(!slot2.hasStack) continue
                val stack2: ItemStack = slot2.stack
                if (stack != stack2 && stack2.item is ItemTool) {
                    val item2 = stack2.item as ItemTool
                    if(ItemUtil.preferredBlock(item2.javaClass) != ItemUtil.preferredBlock(item.javaClass))
                        continue
                    val block = ItemUtil.preferredBlock(item2.javaClass)
                    if (item.getStrVsBlock(stack2, block) == item2.getStrVsBlock(stack, block))
                        return true
                }
            }
        } else if(item is ItemSword) {
            for (i in 0..44) {
                val slot2: Slot = ItemUtil.mc.thePlayer.inventoryContainer.getSlot(i)
                if(!slot2.hasStack) continue
                val comparingStack: ItemStack = slot2.stack
                if (stack != comparingStack && comparingStack.item is ItemSword) {
                    val item2 = comparingStack.item as ItemSword
                    if (item.attackDamage + ItemUtil.getSwordStrength(stack) == item2.attackDamage + ItemUtil.getSwordStrength(comparingStack))
                        return true
                }
            }
        }

        return false
    }

    private fun moveToSlot(fromSlot: Int, toSlot: Int, windowId: Int): Boolean {
        if(toSlot >= 0) {
            if(toSlot < 9) {
                mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, fromSlot, toSlot, 2, mc.thePlayer);
            } else if (toSlot == 69) {
                mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, fromSlot, 0, 1, mc.thePlayer);
            } else {
                mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, fromSlot, 0, 0, mc.thePlayer);
                mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, toSlot, 0, 0, mc.thePlayer);
                mc.playerController.windowClick(mc.thePlayer.inventoryContainer.windowId, fromSlot, 0, 0, mc.thePlayer);
            }
            if (this.enableRandomization.getValue())
                this.time += randomNumber(randomDelay.getValue(), -randomDelay.getValue())
            this.timer.reset()
            return true
        } else return false
    }

    private fun getToolSlot(clazz: Class<*>): Int {
        return if (clazz == ItemPickaxe::class.java)
            this.pickaxeSlot.getValue()
        else if (clazz == ItemAxe::class.java)
            this.axeSlot.getValue()
        else this.shovelSlot.getValue()
    }

    private fun getArmorSlot(armor: ItemArmor): Int {
        // return armor.armorType
        return 69
    }

    private fun getBlocksInInventory(): Int {
        var blockCount = 0
        for (i in 0..44) {
            val slot: Slot = mc.thePlayer.inventoryContainer.getSlot(i)
            if (!slot.hasStack) continue
            val stack = slot.stack
            if(slot.stack.item is ItemBlock)
                blockCount += stack.stackSize
        }
        return blockCount
    }

    private fun slotsAreValid(): Boolean {
        this.slots.forEach { s1 ->
            if(s1.getValue() != 0)
                this.slots.forEach { s2 ->
                    if(s2.getValue() != 0)
                        if (s1.getValue() == s2.getValue() && s1 != s2) return false
                }
        }
        return true
    }

    private fun dropItem(slot: Int, windowId: Int) {
        val slo2t: Slot = ItemUtil.mc.thePlayer.inventoryContainer.getSlot(slot)
        val stack: ItemStack = slo2t.stack
        mc.playerController.windowClick(ItemUtil.mc.thePlayer.inventoryContainer.windowId, slot, 1, 4, ItemUtil.mc.thePlayer)
        if(this.enableRandomization.getValue())
            this.time += randomNumber(randomDelay.getValue(), -randomDelay.getValue())
        this.timer.reset()
    }

    private fun canRun(): Boolean {
        return ((mc.currentScreen is GuiInventory && mc.currentScreen !is GuiContainerCreative) || !openInv.getValue()) && (!player.moving || !whileNotMoving.getValue())
    }
}